### PR TITLE
[SPARK-20305][Spark Core]Master may keep in the state of "COMPELETING…

### PR DESCRIPTION
**Pull request by lvdongr**
## What changes were proposed in this pull request?
Master may keep in the state of "COMPELETING_RECOVERY",then all the application registered cannot get resources, when the leader master change.
This happend when a exception was thrown during the Master trying to recovery(completeRecovery method in the master.scala  ). Then the leader will always in COMPLETING_RECOVERY state ,for the leader can only change to alive from state of RecoveryState.RECOVERING.

## How was this patch tested?
manual tests

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 1]
```diff
@@ -539,7 +539,7 @@ private[deploy] class Master(
 
   private def completeRecovery() {
     // Ensure "only-once" recovery semantics using a short synchronization period.
-    if (state != RecoveryState.RECOVERING) { return }
+    if (state != RecoveryState.RECOVERING && state != RecoveryState.COMPLETING_RECOVERY) { return }
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: @lvdongr would you please explain what will be happened after this change? With your change, which code will call this `completeRecovery`? I see there're 4 places which will call this `completeRecovery`.

Also if we have exception during recovery, should we fail fast?

Comment (lvdongr) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Thank you for your review and suggest. The last change cannot work as I tested today.  I thought completeRecovery would be called again when some workers or drivers response to MasterChanged,  then the master(state is RecoveryState.COMPLETING_RECOVERY) will have the chance to complete the completeRecovery method and change state to ALIVE . I tested , but find no called again after the exception(maybe workers or drivers already response to MasterChanged before the exception).

Comment (lvdongr) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: It seems better to close the master as you say if an exception happened during recovery.So I change the last commit .

**Comment by lvdongr** [comment]
Execute me, Can this issue be closed or threre are some other problem? @jerryshao

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 2]
```diff
@@ -561,6 +561,11 @@ private[deploy] class Master(
     state = RecoveryState.ALIVE
     schedule()
     logInfo("Recovery complete - resuming operations!")
+	} catch {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: At the least, the spelling and spacing in this PR need fixing

Comment (lvdongr) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Thank you very much, I've changed the commit,  you can see  if there are any other problems.

**Comment by jerryshao** [comment]
I'm still not sure what issue you met during recovery and what will be happened if the issue is occurred?

Looks from the fix you provided, what you mainly did is to shutdown rpcEnv, what will be happened after that, will this master process be orphaned or just exited?

**Comment by lvdongr** [comment]
This happend at the time the previous master leader remove the died worker ,clear the worker's node on persistEngine(we use zookeeper),But before the worker node was removed from the zookeeper ,the leader changed. The new master leader recovery from the zookeeper ,and read the died worker node. Then the new leader find the worker died and trying to remove it ,and try to clear the node on zookeeper,but the node has been removed by the previous leader ,so an exception was throw, and the recovery fail. Then the leader will always in COMPLETING_RECOVERY state,and all the application registered cannot get resources .

failfetchresource

**Comment by lvdongr** [comment]
You can see the main method in Master.scala. 

  def main(argStrings: Array[String]) {
    Utils.initDaemon(log)
    val conf = new SparkConf
    val args = new MasterArguments(argStrings, conf)
    val (rpcEnv, _, _) = startRpcEnvAndEndpoint(args.host, args.port, args.webUiPort, conf)
    rpcEnv.awaitTermination()
  }

When the rpcEnv is shut down,the main method will finish,and Master process will stop as I test already. I choose this way ,because the onstop method will be called before stopping master.So the service in master will also be closed,such as webui,metrics,persistenceEngine. I think it will be safer.  Thank you for your last reply @jerryshao

**Comment by jiangxb1987** [comment]
Should we move forward with this PR or should we close this? @jerryshao

**Comment by jerryshao** [comment]
@jiangxb1987 according to what @lvdongr described, seems there's an issue in state transition for recovered master:

> This happend at the time the previous master leader remove the died worker ,clear the worker's node on persistEngine(we use zookeeper),But before the worker node was removed from the zookeeper ,the leader changed. The new master leader recovery from the zookeeper ,and read the died worker node. Then the new leader find the worker died and trying to remove it ,and try to clear the node on zookeeper,but the node has been removed by the previous leader ,so an exception was throw, and the recovery fail. Then the leader will always in COMPLETING_RECOVERY state,and all the application registered cannot get resources .

Based on the description, it looks like a contention issue, but here the fix is just simply a `try catch` solution. IMHO we'd better fix the root cause if we want to move forward this issue.

**Comment by jiangxb1987** [comment]
Agree, let's close this and see whether we can reproduce the root cause of the failure. Thanks! @jerryshao

**Comment by jiangxb1987** [comment]
@lvdongr Would you please close this and submit another PR to address the root cause? Thank you!

**Comment by bzhen** [comment]
hello, has this issue been resolved? was there a follow-up PR?

**Comment by jerzynowakowski** [comment]
Reproduced on spark 2.4.0  - COMPELETING_RECOVERY state was inifinte. Manual kill of sparks helped only.

**Pull request closed without merging**